Race condition in flightcontrol causing occasional test failure

When running unrelated tests, I saw a one-off failure in `TestCancelBoth` of `util/flightcontrol/flightcontrol_test.go`:
```
=== CONT  TestCancelBoth
    flightcontrol_test.go:199:
                Error Trace:    flightcontrol_test.go:199
                Error:          Not equal:
                                expected: "bar"
                                actual  : "baz"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -bar
                                +baz
                Test:           TestCancelBoth
    flightcontrol_test.go:204:
                Error Trace:    flightcontrol_test.go:204
                Error:          Not equal:
                                expected: 3
                                actual  : 4
                Test:           TestCancelBoth
--- FAIL: TestCancelBoth (0.45s)
```

It doesn't reproduce very often, but I found if I cd into `util/flightcontrol` and run `go test -run TestCancelBoth -count 250` I'll almost always get at least 1 failure (same as the one above, it's always the same errors on the same lines). Others might need to increase `-count` depending on the hardware being run on I suppose.

Not 100% sure but I think it happens when:
1. `g.Do` is called [here](https://github.com/moby/buildkit/blob/b2195cec759428a7775a6fe01799b961d14a81bc/util/flightcontrol/flightcontrol_test.go#L192). It returns, but the go routine it created internally is still [here and hasn't been scheduled to run yet](https://github.com/moby/buildkit/blob/b2195cec759428a7775a6fe01799b961d14a81bc/util/flightcontrol/flightcontrol.go#L71).
2. This second `g.Do` gets called [here](https://github.com/moby/buildkit/blob/b2195cec759428a7775a6fe01799b961d14a81bc/util/flightcontrol/flightcontrol_test.go#L197) (with different func as callback). It gets all the way to [here](https://github.com/moby/buildkit/blob/b2195cec759428a7775a6fe01799b961d14a81bc/util/flightcontrol/flightcontrol.go#L143) and thus returns the value from the previous `g.Do` call, even though the callback func is different.

@tonistiigi If the above is correct then I think this was introduced in [this commit](https://github.com/moby/buildkit/pull/2195). What do you think the right behavior is here? Should it be that once a call to `g.Do` returns then any state for the key is completely removed? Or do you think that would undo the improvements to go routine contention from the previous commit?

--- a/cache/blobs.go
+++ b/cache/blobs.go
@@ -2,6 +2,7 @@ package cache
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/containerd/containerd/diff"
 	"github.com/containerd/containerd/leases"
@@ -44,29 +45,19 @@ func (sr *immutableRef) computeBlobChain(ctx context.Context, createIfNeeded boo
 }
 
 func computeBlobChain(ctx context.Context, sr *immutableRef, createIfNeeded bool, compressionType compression.Type, forceCompression bool, s session.Group) error {
-	baseCtx := ctx
 	eg, ctx := errgroup.WithContext(ctx)
-	var currentDescr ocispecs.Descriptor
 	if sr.parent != nil {
 		eg.Go(func() error {
 			return computeBlobChain(ctx, sr.parent, createIfNeeded, compressionType, forceCompression, s)
 		})
 	}
+
 	eg.Go(func() error {
-		dp, err := g.Do(ctx, sr.ID(), func(ctx context.Context) (interface{}, error) {
-			refInfo := sr.Info()
-			if refInfo.Blob != "" {
-				if forceCompression {
-					desc, err := sr.ociDesc()
-					if err != nil {
-						return nil, err
-					}
-					if err := ensureCompression(ctx, sr, desc, compressionType, s); err != nil {
-						return nil, err
-					}
-				}
-				return nil, nil
-			} else if !createIfNeeded {
+		v, err := g.Do(ctx, fmt.Sprintf("%s-%t", sr.ID(), createIfNeeded), func(ctx context.Context) (interface{}, error) {
+			if getBlob(sr.md) != "" {
+				return sr.ociDesc()
+			}
+			if !createIfNeeded {
 				return nil, errors.WithStack(ErrNoBlobs)
 			}
 
@@ -80,91 +71,83 @@ func computeBlobChain(ctx context.Context, sr *immutableRef, createIfNeeded bool
 				return nil, errors.Errorf("unknown layer compression type: %q", compressionType)
 			}
 
-			var descr ocispecs.Descriptor
-			var err error
-
-			if descr.Digest == "" {
-				// reference needs to be committed
-				var lower []mount.Mount
-				if sr.parent != nil {
-					m, err := sr.parent.Mount(ctx, true, s)
-					if err != nil {
-						return nil, err
-					}
-					var release func() error
-					lower, release, err = m.Mount()
-					if err != nil {
-						return nil, err
-					}
-					if release != nil {
-						defer release()
-					}
-				}
-				m, err := sr.Mount(ctx, true, s)
+			var lower []mount.Mount
+			if sr.parent != nil {
+				m, err := sr.parent.Mount(ctx, true, s)
 				if err != nil {
 					return nil, err
 				}
-				upper, release, err := m.Mount()
+				var release func() error
+				lower, release, err = m.Mount()
 				if err != nil {
 					return nil, err
 				}
 				if release != nil {
 					defer release()
 				}
-				descr, err = sr.cm.Differ.Compare(ctx, lower, upper,
-					diff.WithMediaType(mediaType),
-					diff.WithReference(sr.ID()),
-				)
-				if err != nil {
-					return nil, err
-				}
+			}
+			m, err := sr.Mount(ctx, true, s)
+			if err != nil {
+				return nil, err
+			}
+			upper, release, err := m.Mount()
+			if err != nil {
+				return nil, err
+			}
+			if release != nil {
+				defer release()
+			}
+			desc, err := sr.cm.Differ.Compare(ctx, lower, upper,
+				diff.WithMediaType(mediaType),
+				diff.WithReference(sr.ID()),
+			)
+			if err != nil {
+				return nil, err
 			}
 
-			if descr.Annotations == nil {
-				descr.Annotations = map[string]string{}
+			if desc.Annotations == nil {
+				desc.Annotations = map[string]string{}
 			}
 
-			info, err := sr.cm.ContentStore.Info(ctx, descr.Digest)
+			info, err := sr.cm.ContentStore.Info(ctx, desc.Digest)
 			if err != nil {
 				return nil, err
 			}
 
 			if diffID, ok := info.Labels[containerdUncompressed]; ok {
-				descr.Annotations[containerdUncompressed] = diffID
-			} else if compressionType == compression.Uncompressed {
-				descr.Annotations[containerdUncompressed] = descr.Digest.String()
+				desc.Annotations[containerdUncompressed] = diffID
+			} else if mediaType == ocispecs.MediaTypeImageLayer {
+				desc.Annotations[containerdUncompressed] = desc.Digest.String()
 			} else {
 				return nil, errors.Errorf("unknown layer compression type")
 			}
 
-			if forceCompression {
-				if err := ensureCompression(ctx, sr, descr, compressionType, s); err != nil {
-					return nil, err
-				}
+			if err := sr.setBlob(ctx, desc); err != nil {
+				return nil, err
 			}
 
-			return descr, nil
-
+			return desc, nil
 		})
 		if err != nil {
 			return err
 		}
+		descr, ok := v.(ocispecs.Descriptor)
+		if !ok {
+			return fmt.Errorf("invalid descriptor returned by differ while computing blob for %s", sr.ID())
+		}
 
-		if dp != nil {
-			currentDescr = dp.(ocispecs.Descriptor)
+		if forceCompression {
+			if err := ensureCompression(ctx, sr, descr, compressionType, s); err != nil {
+				return err
+			}
 		}
 		return nil
 	})
-	err := eg.Wait()
-	if err != nil {
+
+	if err := eg.Wait(); err != nil {
 		return err
 	}
-	if currentDescr.Digest != "" {
-		if err := sr.setBlob(baseCtx, currentDescr); err != nil {
-			return err
-		}
-	}
-	return nil
+	return sr.setChains(ctx)
 }
 
 // setBlob associates a blob with the cache record.
@@ -183,27 +166,20 @@ func (sr *immutableRef) setBlob(ctx context.Context, desc ocispecs.Descriptor) e
 		return err
 	}
 
+	compressionType := compression.FromMediaType(desc.MediaType)
+	if compressionType == compression.UnknownCompression {
+		return errors.Errorf("unhandled layer media type: %q", desc.MediaType)
+	}
+
 	sr.mu.Lock()
 	defer sr.mu.Unlock()
 
-	if getChainID(sr.md) != "" {
+	if getBlob(sr.md) != "" {
 		return nil
 	}
 
 	if err := sr.finalize(ctx); err != nil {
 		return err
-	}
-
-	p := sr.parent
-	var parentChainID digest.Digest
-	var parentBlobChainID digest.Digest
-	if p != nil {
-		pInfo := p.Info()
-		if pInfo.ChainID == "" || pInfo.BlobChainID == "" {
-			return errors.Errorf("failed to set blob for reference with non-addressable parent")
-		}
-		parentChainID = pInfo.ChainID
-		parentBlobChainID = pInfo.BlobChainID
 	}
 
 	if err := sr.cm.LeaseManager.AddResource(ctx, leases.Lease{ID: sr.ID()}, leases.Resource{
@@ -215,16 +191,45 @@ func (sr *immutableRef) setBlob(ctx context.Context, desc ocispecs.Descriptor) e
 
 	queueDiffID(sr.md, diffID.String())
 	queueBlob(sr.md, desc.Digest.String())
-	chainID := diffID
-	blobChainID := imagespecidentity.ChainID([]digest.Digest{desc.Digest, diffID})
-	if parentChainID != "" {
-		chainID = imagespecidentity.ChainID([]digest.Digest{parentChainID, chainID})
-		blobChainID = imagespecidentity.ChainID([]digest.Digest{parentBlobChainID, blobChainID})
-	}
-	queueChainID(sr.md, chainID.String())
-	queueBlobChainID(sr.md, blobChainID.String())
 	queueMediaType(sr.md, desc.MediaType)
 	queueBlobSize(sr.md, desc.Size)
+	if err := sr.md.Commit(); err != nil {
+		return err
+	}
+
+	if err := sr.addCompressionBlob(ctx, desc.Digest, compressionType); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (sr *immutableRef) setChains(ctx context.Context) error {
+	if _, ok := leases.FromContext(ctx); !ok {
+		return errors.Errorf("missing lease requirement for setChains")
+	}
+
+	sr.mu.Lock()
+	defer sr.mu.Unlock()
+
+	if getChainID(sr.md) != "" {
+		return nil
+	}
+
+	var chainIDs []digest.Digest
+	var blobChainIDs []digest.Digest
+	if sr.parent != nil {
+		chainIDs = append(chainIDs, digest.Digest(getChainID(sr.parent.md)))
+		blobChainIDs = append(blobChainIDs, digest.Digest(getBlobChainID(sr.parent.md)))
+	}
+	diffID := digest.Digest(getDiffID(sr.md))
+	chainIDs = append(chainIDs, diffID)
+	blobChainIDs = append(blobChainIDs, imagespecidentity.ChainID([]digest.Digest{digest.Digest(getBlob(sr.md)), diffID}))
+
+	chainID := imagespecidentity.ChainID(chainIDs)
+	blobChainID := imagespecidentity.ChainID(blobChainIDs)
+
+	queueChainID(sr.md, chainID.String())
+	queueBlobChainID(sr.md, blobChainID.String())
 	if err := sr.md.Commit(); err != nil {
 		return err
 	}
@@ -243,36 +248,39 @@ func isTypeWindows(sr *immutableRef) bool {
 
 // ensureCompression ensures the specified ref has the blob of the specified compression Type.
 func ensureCompression(ctx context.Context, ref *immutableRef, desc ocispecs.Descriptor, compressionType compression.Type, s session.Group) error {
-	// Resolve converters
-	layerConvertFunc, _, err := getConverters(desc, compressionType)
-	if err != nil {
-		return err
-	} else if layerConvertFunc == nil {
-		return nil // no need to convert
-	}
+	_, err := g.Do(ctx, fmt.Sprintf("%s-%d", desc.Digest, compressionType), func(ctx context.Context) (interface{}, error) {
+		// Resolve converters
+		layerConvertFunc, _, err := getConverters(desc, compressionType)
+		if err != nil {
+			return nil, err
+		} else if layerConvertFunc == nil {
+			return nil, nil // no need to convert
+		}
 
-	// First, lookup local content store
-	if _, err := ref.getCompressionBlob(ctx, compressionType); err == nil {
-		return nil // found the compression variant. no need to convert.
-	}
+		// First, lookup local content store
+		if _, err := ref.getCompressionBlob(ctx, compressionType); err == nil {
+			return nil, nil // found the compression variant. no need to convert.
+		}
 
-	// Convert layer compression type
-	if err := (lazyRefProvider{
-		ref:     ref,
-		desc:    desc,
-		dh:      ref.descHandlers[desc.Digest],
-		session: s,
-	}).Unlazy(ctx); err != nil {
-		return err
-	}
-	newDesc, err := layerConvertFunc(ctx, ref.cm.ContentStore, desc)
-	if err != nil {
-		return err
-	}
+		// Convert layer compression type
+		if err := (lazyRefProvider{
+			ref:     ref,
+			desc:    desc,
+			dh:      ref.descHandlers[desc.Digest],
+			session: s,
+		}).Unlazy(ctx); err != nil {
+			return nil, err
+		}
+		newDesc, err := layerConvertFunc(ctx, ref.cm.ContentStore, desc)
+		if err != nil {
+			return nil, err
+		}
 
-	// Start to track converted layer
-	if err := ref.addCompressionBlob(ctx, newDesc.Digest, compressionType); err != nil {
-		return err
-	}
-	return nil
+		// Start to track converted layer
+		if err := ref.addCompressionBlob(ctx, newDesc.Digest, compressionType); err != nil {
+			return nil, err
+		}
+		return nil, nil
+	})
+	return err
 }

--- a/cache/manager_test.go
+++ b/cache/manager_test.go
@@ -5,16 +5,20 @@ import (
 	"bytes"
 	"compress/gzip"
 	"context"
+	"fmt"
 	"io"
 	"io/ioutil"
 	"os"
 	"path/filepath"
 	"runtime"
+	"strconv"
 	"testing"
+	"time"
 
 	"github.com/containerd/containerd/content"
 	"github.com/containerd/containerd/content/local"
 	"github.com/containerd/containerd/diff/apply"
+	"github.com/containerd/containerd/diff/walking"
 	"github.com/containerd/containerd/leases"
 	ctdmetadata "github.com/containerd/containerd/metadata"
 	"github.com/containerd/containerd/namespaces"
@@ -22,14 +26,19 @@ import (
 	"github.com/containerd/containerd/snapshots/native"
 	"github.com/moby/buildkit/cache/metadata"
 	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/session"
 	"github.com/moby/buildkit/snapshot"
 	containerdsnapshot "github.com/moby/buildkit/snapshot/containerd"
+	"github.com/moby/buildkit/util/compression"
+	"github.com/moby/buildkit/util/contentutil"
 	"github.com/moby/buildkit/util/leaseutil"
+	"github.com/moby/buildkit/util/winlayers"
 	digest "github.com/opencontainers/go-digest"
 	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/pkg/errors"
 	"github.com/stretchr/testify/require"
 	bolt "go.etcd.io/bbolt"
+	"golang.org/x/sync/errgroup"
 )
 
 type cmOpt struct {
@@ -116,15 +125,17 @@ func newCacheManager(ctx context.Context, opt cmOpt) (co *cmOut, cleanup func() 
 		return nil, nil, err
 	}
 
-	lm := ctdmetadata.NewLeaseManager(mdb)
+	store = containerdsnapshot.NewContentStore(mdb.ContentStore(), ns)
+	lm := leaseutil.WithNamespace(ctdmetadata.NewLeaseManager(mdb), ns)
 
 	cm, err := NewManager(ManagerOpt{
 		Snapshotter:    snapshot.FromContainerdSnapshotter(opt.snapshotterName, containerdsnapshot.NSSnapshotter(ns, mdb.Snapshotter(opt.snapshotterName)), nil),
 		MetadataStore:  md,
-		ContentStore:   mdb.ContentStore(),
-		LeaseManager:   leaseutil.WithNamespace(lm, ns),
+		ContentStore:   store,
+		LeaseManager:   lm,
 		GarbageCollect: mdb.GarbageCollect,
-		Applier:        apply.NewFileSystemApplier(mdb.ContentStore()),
+		Applier:        winlayers.NewFileSystemApplierWithWindows(store, apply.NewFileSystemApplier(store)),
+		Differ:         winlayers.NewWalkingDiffWithWindows(store, walking.NewWalkingDiff(store)),
 	})
 	if err != nil {
 		return nil, nil, err
@@ -132,7 +143,7 @@ func newCacheManager(ctx context.Context, opt cmOpt) (co *cmOut, cleanup func() 
 	return &cmOut{
 		manager: cm,
 		lm:      lm,
-		cs:      mdb.ContentStore(),
+		cs:      store,
 	}, cleanup, nil
 }
 
@@ -506,6 +517,8 @@ func TestExtractOnMutable(t *testing.T) {
 	err = snap.(*immutableRef).setBlob(leaseCtx, desc)
 	done(context.TODO())
 	require.NoError(t, err)
+	err = snap.(*immutableRef).setChains(leaseCtx)
+	require.NoError(t, err)
 
 	snap2, err := cm.GetByBlob(ctx, desc2, snap)
 	require.NoError(t, err)
@@ -623,6 +636,8 @@ func TestSetBlob(t *testing.T) {
 
 	err = snap.(*immutableRef).setBlob(ctx, desc)
 	require.NoError(t, err)
+	err = snap.(*immutableRef).setChains(ctx)
+	require.NoError(t, err)
 
 	info = snap.Info()
 	require.Equal(t, desc.Annotations["containerd.io/uncompressed"], string(info.DiffID))
@@ -646,6 +661,8 @@ func TestSetBlob(t *testing.T) {
 	require.NoError(t, err)
 
 	err = snap2.(*immutableRef).setBlob(ctx, desc2)
+	require.NoError(t, err)
+	err = snap2.(*immutableRef).setChains(ctx)
 	require.NoError(t, err)
 
 	info2 := snap2.Info()
@@ -997,6 +1014,171 @@ func TestLazyCommit(t *testing.T) {
 	require.Equal(t, true, errors.Is(err, errNotFound))
 }
 
+func TestGetRemote(t *testing.T) {
+	t.Parallel()
+	// windows fails when lazy blob is being extracted with "invalid windows mount type: 'bind'"
+	if runtime.GOOS != "linux" {
+		t.Skipf("unsupported GOOS: %s", runtime.GOOS)
+	}
+
+	ctx := namespaces.WithNamespace(context.Background(), "buildkit-test")
+
+	tmpdir, err := ioutil.TempDir("", "cachemanager")
+	require.NoError(t, err)
+	defer os.RemoveAll(tmpdir)
+
+	snapshotter, err := native.NewSnapshotter(filepath.Join(tmpdir, "snapshots"))
+	require.NoError(t, err)
+
+	co, cleanup, err := newCacheManager(ctx, cmOpt{
+		snapshotter:     snapshotter,
+		snapshotterName: "native",
+	})
+	require.NoError(t, err)
+	defer cleanup()
+	cm := co.manager
+
+	ctx, done, err := leaseutil.WithLease(ctx, co.lm, leaseutil.MakeTemporary)
+	require.NoError(t, err)
+	defer done(context.TODO())
+
+	contentBuffer := contentutil.NewBuffer()
+
+	descHandlers := DescHandlers(map[digest.Digest]*DescHandler{})
+
+	// make some lazy refs from blobs
+	expectedContent := map[digest.Digest]struct{}{}
+	var descs []ocispecs.Descriptor
+	for i := 0; i < 2; i++ {
+		blobmap := map[string]string{"foo": strconv.Itoa(i)}
+		blobBytes, desc, err := mapToBlob(blobmap, true)
+		require.NoError(t, err)
+
+		expectedContent[desc.Digest] = struct{}{}
+		descs = append(descs, desc)
+
+		cw, err := contentBuffer.Writer(ctx)
+		require.NoError(t, err)
+		_, err = cw.Write(blobBytes)
+		require.NoError(t, err)
+		err = cw.Commit(ctx, 0, cw.Digest())
+		require.NoError(t, err)
+
+		descHandlers[desc.Digest] = &DescHandler{
+			Provider: func(_ session.Group) content.Provider { return contentBuffer },
+		}
+
+		_, uncompressedDesc, err := mapToBlob(blobmap, false)
+		require.NoError(t, err)
+		expectedContent[uncompressedDesc.Digest] = struct{}{}
+	}
+
+	// Create 3 levels of mutable refs, where each parent ref has 2 children (this tests parallel creation of
+	// overlapping blob chains).
+	lazyRef, err := cm.GetByBlob(ctx, descs[0], nil, descHandlers)
+	require.NoError(t, err)
+
+	refs := []ImmutableRef{lazyRef}
+	for i := 0; i < 3; i++ {
+		var newRefs []ImmutableRef
+		for j, ir := range refs {
+			for k := 0; k < 2; k++ {
+				mutRef, err := cm.New(ctx, ir, nil, descHandlers)
+				require.NoError(t, err)
+
+				m, err := mutRef.Mount(ctx, false, nil)
+				require.NoError(t, err)
+
+				lm := snapshot.LocalMounter(m)
+				target, err := lm.Mount()
+				require.NoError(t, err)
+
+				f, err := os.Create(filepath.Join(target, fmt.Sprintf("%d-%d-%d", i, j, k)))
+				require.NoError(t, err)
+				err = os.Chtimes(f.Name(), time.Unix(0, 0), time.Unix(0, 0))
+				require.NoError(t, err)
+
+				_, desc, err := fileToBlob(f, true)
+				require.NoError(t, err)
+				expectedContent[desc.Digest] = struct{}{}
+				_, desc, err = fileToBlob(f, false)
+				require.NoError(t, err)
+				expectedContent[desc.Digest] = struct{}{}
+
+				f.Close()
+				err = lm.Unmount()
+				require.NoError(t, err)
+
+				immutRef, err := mutRef.Commit(ctx)
+				require.NoError(t, err)
+				newRefs = append(newRefs, immutRef)
+			}
+		}
+		refs = newRefs
+	}
+
+	// also test the original lazyRef to get coverage for refs that don't have to be extracted from the snapshotter
+	lazyRef2, err := cm.GetByBlob(ctx, descs[1], nil, descHandlers)
+	require.NoError(t, err)
+	refs = append(refs, lazyRef2)
+
+	checkNumBlobs(ctx, t, co.cs, 1)
+
+	// Call GetRemote on all the refs
+	eg, egctx := errgroup.WithContext(ctx)
+	for _, ir := range refs {
+		ir := ir.(*immutableRef)
+		for _, compressionType := range []compression.Type{compression.Uncompressed, compression.Gzip} {
+			compressionType := compressionType
+			eg.Go(func() error {
+				remote, err := ir.GetRemote(egctx, true, compressionType, true, nil)
+				require.NoError(t, err)
+				refChain := ir.parentRefChain()
+				for i, desc := range remote.Descriptors {
+					switch compressionType {
+					case compression.Uncompressed:
+						require.Equal(t, ocispecs.MediaTypeImageLayer, desc.MediaType)
+					case compression.Gzip:
+						require.Equal(t, ocispecs.MediaTypeImageLayerGzip, desc.MediaType)
+					default:
+						require.Fail(t, "unhandled media type", compressionType)
+					}
+					require.Contains(t, expectedContent, desc.Digest)
+
+					r := refChain[i]
+					isLazy, err := r.isLazy(egctx)
+					require.NoError(t, err)
+					info, err := r.getCompressionBlob(egctx, compressionType)
+					if isLazy {
+						require.Error(t, err)
+					} else {
+						require.NoError(t, err)
+						require.Equal(t, info.Digest, desc.Digest)
+					}
+				}
+				return nil
+			})
+		}
+	}
+	require.NoError(t, eg.Wait())
+
+	// verify there's a 1-to-1 mapping between the content store and what we expected to be there
+	err = co.cs.Walk(ctx, func(info content.Info) error {
+		var matched bool
+		for expected := range expectedContent {
+			if info.Digest == expected {
+				delete(expectedContent, expected)
+				matched = true
+				break
+			}
+		}
+		require.True(t, matched, "match for blob: %s", info.Digest)
+		return nil
+	})
+	require.NoError(t, err)
+	require.Equal(t, map[digest.Digest]struct{}{}, expectedContent)
+}
+
 func checkDiskUsage(ctx context.Context, t *testing.T, cm Manager, inuse, unused int) {
 	du, err := cm.DiskUsage(ctx, client.DiskUsageInfo{})
 	require.NoError(t, err)
@@ -1073,6 +1255,58 @@ func mapToBlob(m map[string]string, compress bool) ([]byte, ocispecs.Descriptor,
 			return nil, ocispecs.Descriptor{}, err
 		}
 	}
+	if err := tw.Close(); err != nil {
+		return nil, ocispecs.Descriptor{}, err
+	}
+	if err := dest.Close(); err != nil {
+		return nil, ocispecs.Descriptor{}, err
+	}
+
+	mediaType := ocispecs.MediaTypeImageLayer
+	if compress {
+		mediaType = ocispecs.MediaTypeImageLayerGzip
+	}
+	return buf.Bytes(), ocispecs.Descriptor{
+		Digest:    digest.FromBytes(buf.Bytes()),
+		MediaType: mediaType,
+		Size:      int64(buf.Len()),
+		Annotations: map[string]string{
+			"containerd.io/uncompressed": sha.Digest().String(),
+		},
+	}, nil
+}
+
+func fileToBlob(file *os.File, compress bool) ([]byte, ocispecs.Descriptor, error) {
+	buf := bytes.NewBuffer(nil)
+	sha := digest.SHA256.Digester()
+
+	var dest io.WriteCloser = bufferCloser{buf}
+	if compress {
+		dest = gzip.NewWriter(buf)
+	}
+	tw := tar.NewWriter(io.MultiWriter(sha.Hash(), dest))
+
+	info, err := file.Stat()
+	if err != nil {
+		return nil, ocispecs.Descriptor{}, err
+	}
+
+	fi, err := tar.FileInfoHeader(info, "")
+	if err != nil {
+		return nil, ocispecs.Descriptor{}, err
+	}
+	fi.Format = tar.FormatPAX
+	fi.ModTime = fi.ModTime.Truncate(time.Second)
+	fi.AccessTime = time.Time{}
+	fi.ChangeTime = time.Time{}
+
+	if err := tw.WriteHeader(fi); err != nil {
+		return nil, ocispecs.Descriptor{}, err
+	}
+	if _, err := io.Copy(tw, file); err != nil {
+		return nil, ocispecs.Descriptor{}, err
+	}
+
 	if err := tw.Close(); err != nil {
 		return nil, ocispecs.Descriptor{}, err
 	}

--- a/cache/refs.go
+++ b/cache/refs.go
@@ -403,7 +403,11 @@ func (sr *immutableRef) getCompressionBlob(ctx context.Context, compressionType 
 		if err != nil {
 			return content.Info{}, err
 		}
-		return cs.Info(ctx, dgst)
+		info, err := cs.Info(ctx, dgst)
+		if err != nil {
+			return content.Info{}, err
+		}
+		return info, nil
 	}
 	return content.Info{}, errdefs.ErrNotFound
 }

--- a/cache/remote.go
+++ b/cache/remote.go
@@ -224,7 +224,16 @@ func (p lazyRefProvider) Unlazy(ctx context.Context) error {
 				}
 			}
 		}
-		return nil, err
+
+		compressionType := compression.FromMediaType(p.desc.MediaType)
+		if compressionType == compression.UnknownCompression {
+			return nil, errors.Errorf("unhandled layer media type: %q", p.desc.MediaType)
+		}
+
+		if err := p.ref.addCompressionBlob(ctx, p.desc.Digest, compressionType); err != nil {
+			return nil, err
+		}
+		return nil, nil
 	})
 	return err
 }

--- a/solver/jobs.go
+++ b/solver/jobs.go
@@ -715,7 +715,7 @@ func (s *sharedOp) CacheMap(ctx context.Context, index int) (resp *cacheMapResp,
 	if err != nil {
 		return nil, err
 	}
-	res, err := s.g.Do(ctx, "cachemap", func(ctx context.Context) (ret interface{}, retErr error) {
+	res, err := s.g.Do(ctx, fmt.Sprintf("cachemap-%d", index), func(ctx context.Context) (ret interface{}, retErr error) {
 		if s.cacheRes != nil && s.cacheDone || index < len(s.cacheRes) {
 			return s.cacheRes, nil
 		}

--- a/source/containerimage/pull.go
+++ b/source/containerimage/pull.go
@@ -83,6 +83,7 @@ func (is *Source) ResolveImageConfig(ctx context.Context, ref string, opt llb.Re
 	if err != nil {
 		return "", nil, err
 	}
+	key += rm.String()
 
 	res, err := is.g.Do(ctx, key, func(ctx context.Context) (interface{}, error) {
 		res := resolver.DefaultPool.GetResolver(is.RegistryHosts, ref, "pull", sm, g).WithImageStore(is.ImageStore, rm)

--- a/util/compression/compression.go
+++ b/util/compression/compression.go
@@ -40,6 +40,17 @@ func (ct Type) String() string {
 	}
 }
 
+func FromMediaType(mediaType string) Type {
+	switch toOCILayerType[mediaType] {
+	case ocispecs.MediaTypeImageLayer:
+		return Uncompressed
+	case ocispecs.MediaTypeImageLayerGzip:
+		return Gzip
+	default:
+		return UnknownCompression
+	}
+}
+
 // DetectLayerMediaType returns media type from existing blob data.
 func DetectLayerMediaType(ctx context.Context, cs content.Store, id digest.Digest, oci bool) (string, error) {
 	ra, err := cs.ReaderAt(ctx, ocispecs.Descriptor{Digest: id})

--- a/util/flightcontrol/flightcontrol_test.go
+++ b/util/flightcontrol/flightcontrol_test.go
@@ -193,15 +193,11 @@ func TestCancelBoth(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, ret1, "bar")
 
-	f2 := testFunc(100*time.Millisecond, "baz", &counter)
-	ret1, err = g.Do(context.TODO(), "foo", f2)
-	assert.NoError(t, err)
-	assert.Equal(t, ret1, "baz")
 	ret1, err = g.Do(context.TODO(), "abc", f)
 	assert.NoError(t, err)
 	assert.Equal(t, ret1, "bar")
 
-	assert.Equal(t, counter, int64(4))
+	assert.Equal(t, counter, int64(3))
 }
 
 func TestContention(t *testing.T) {

